Bump operator sdk v1.13.0

AKamyshnikova opened the pull request:
| Q               | A
| --------------- | ---
| Bug fix?        | []
| New feature?    | [x]
| API breaks?     | []
| Deprecations?   | []
| Related tickets | fixes #X, partially #Y, mentioned in #Z
| License         | Apache 2.0


### What's in this PR?
Use latest operator-sdk version.

Fixes #351 


### Additional context
v1.x.x of operator-sdk contains several breaking changes and require massive refactor of operator structure.
https://sdk.operatorframework.io/docs/building-operators/golang/migration/


### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x ] Implementation tested
- [ ] Logging code meets the guideline
- [ ] User guide and development docs updated (if needed)
- [ ] Append changelog

### To Do
Work is still in progress.
Multi-casskop is not finally switched yet.

AKamyshnikova committed:
[operator-sdk-update] Refactor structure

Upgrade to operator-sdk 1.x version requires major structure
refactoring [1].
Current commit updates file structure.

[1] - https://sdk.operatorframework.io/docs/building-operators/golang/migration/

Signed-off-by: akamyshnikova <akamyshnikova@mirantis.com>

AKamyshnikova committed:
Bump operator-sdk v1.13.0

Refactor pkg/apis and pkg/controller folders

Signed-off-by: akamyshnikova <akamyshnikova@mirantis.com>

AKamyshnikova commented:
@cscetbon Hi! Could you upload newer version of casskop-build with v1.13.0 tag to orangeopensource? So I can update circleci settings as well.

cscetbon commented:
> @cscetbon Hi! Could you upload newer version of casskop-build with v1.13.0 tag to orangeopensource? So I can update circleci settings as well.

@AKamyshnikova  Done ✅

AKamyshnikova commented:
@cscetbon Thanks!
I faced issues with update-crds - it passes locally, but fails in CI...
Why we need such workaround for updating CRDs? 
Seems that v1alpha1 folders has been renamed to v2 and it broke compatibility, so we need to leave v1alpha1 folders for backward compatibility. Was this considered?

cscetbon commented:
That was for existing clusters. If you have to get rid of v1 do it

AKamyshnikova commented:
@cscetbon 
I just a little bit confused, I don't get any diff after running "make generate" locally. circleci/Dockerfile uses older controller-gen version, may be this is the problem. Sorry to bother, could you rebuild and upload casskop-build with v1.13.0 tag one more time? I've updated controller-gen version in latest commit.

cscetbon commented:
@AKamyshnikova I just did it again 😉

AKamyshnikova commented:
@cscetbon This pull request is ready for review.

cscetbon commented:
@AKamyshnikova it seems to be a big work. Not sure why kustomize is used but I started to review your PR, not done yet. But e2e tests are failing and you can fix them in the meantime ?

AKamyshnikova commented:
@cscetbon Yeah, will fix this. Kustomize is part of new things with v1.x https://sdk.operatorframework.io/docs/building-operators/golang/migration/#what-is-new

cscetbon commented:
@AKamyshnikova lmk when it's ready and I'll rerun the tests. You can try them locally using k3d like circle does

cscetbon commented on controllers/suite_test.go:
```diff
@@ -0,0 +1,71 @@
+package controllers
```
what's that file for ? future tests ?

cscetbon commented on multi-casskop/api/v2/cassandramulticluster_types.go:
```diff
@@ -11,7 +11,7 @@ import (
 // +k8s:openapi-gen=true
 type MultiCasskopSpec struct {
-	DeleteCassandraCluster *bool                             `json:"deleteCassandraCluster,omitempty"`
+	DeleteCassandraCluster *bool                       `json:"deleteCassandraCluster,omitempty"`
```
indentation with the next lines

cscetbon reviewed (changes requested):
@AKamyshnikova a few comments. I like what you did. I would probably have done the renaming in a different PR to keep that one small but it's okay. thanks for your hard work !

AKamyshnikova commented on .circleci/config.yml:
```diff
@@ -114,6 +114,7 @@ orbs:
               command: |
                 GO111MODULE=on go install github.com/mikefarah/yq/v4@latest
                 make generate
+                git status
```
yes, will drop this

AKamyshnikova commented on Dockerfile:
```diff
@@ -8,16 +8,35 @@ RUN mkdir -p /tmp && chown casskop /tmp
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+COPY vendor/ vendor/
```
If I did not specify this, I'm getting
#13 0.331 go mod download: no modules specified (see 'go help mod download')
#13 ERROR: executor failed running [/bin/sh -c go mod download]: exit code: 1

AKamyshnikova commented on Makefile:
```diff
@@ -89,21 +89,20 @@ SPEC_PROPS = $(FIRST_VERSION).schema.openAPIV3Schema.properties.spec.properties
+         sed -e '1,/versions/d' -e "1,$${end}s/^..//" $$crd >> /tmp/$$crdname; \
+         cp /tmp/$$crdname $$crd; \
+         yq -i e '$(FIRST_VERSION).storage = false' $$crd; \
+    done
```
Strange indentation seems correct in my IDE, will fix this.

AKamyshnikova commented on api/v2/common.go:
```diff
@@ -1,4 +1,4 @@
-package common
+package v2
```
Will reverse this

cscetbon commented:
@AKamyshnikova let me know when your branch is ready but atm unit tests are failing.

cscetbon commented:
@AKamyshnikova trying to rerun tests on your branch I see that TestCassandraBackupAlreadyExists is failing. let me know when you think it's ready

AKamyshnikova commented:
@cscetbon I've updated client-go to v.22.0 and controller-runtime to v0.9.0 - it seems to break unittests. I reverted to use v.0.19.3 and v0.6.5, I think update of these dependencies can be done in separate pull request as requires some additional work. Now tests are passing.

cscetbon commented:
👏  @AKamyshnikova unit tests are fixed now. e2e tests are failing though because of an error with helm
```
go: unknown environment setting GO111MODULE=true
helm install ******* helm/cassandra-operator --set image.tag=trigger-integration
Error: INSTALLATION FAILED: failed to install CRD crds/*.yaml: no objects visited
make: *** [Makefile:279: kuttl-test-fix-arg] Error 1
```

AKamyshnikova commented:
@cscetbon Seems I accidentally created file helm/cassandra-operator/crds/*.yaml :(  I dropped it now.

cscetbon commented:
Helm install doesn't fail anymore but e2e are failing. It seems it can't find the status key in the object for some reason

AKamyshnikova commented:
@cscetbon One of changes is that ready is dropped https://sdk.operatorframework.io/docs/upgrading-sdk-version/v1.0.0/#removed-package-pkgready I missed that helm templates require update for this, fixed now.

cscetbon commented:
e2e Tests ran again. Still missing status key in k8s object

AKamyshnikova commented:
> e2e Tests ran again. Still missing status key in k8s object

I run tests locally and don't get missing .status I'm not sure how to debug this for now.

AKamyshnikova commented:
@cscetbon Hi! Could you allow running rest of the checks? I dropped +kubebuilder:subresource:status in latest patch set, may be this was a problem.

cscetbon commented:
> @cscetbon Hi! Could you allow running rest of the checks? I dropped +kubebuilder:subresource:status in latest patch set, may be this was a problem. 

Better now. There is one remaining task failing but I'm not sure it's on you this time. I'm gonna review it. If in the meantime you think the failing test is on you don't hesitate to fix it 😉

AKamyshnikova commented:
I checked and was some strange diff in  two website/static/slides*.js files, hopefully fixed now.

cscetbon reviewed (changes requested):
Thanks for the hard work. A few comments and changes asked. Also do you know what would be the impact for a person who's already running the current version to move to this one ?

AKamyshnikova commented:
> Thanks for the hard work. A few comments and changes asked. Also do you know what would be the impact for a person who's already running the current version to move to this one ?

While I was testing this pull request I've updated my test cluster (which use casskop v2.0.2  and has cassandra clusters) with this version and things work in the same way. So, from user perspective it would be just update to latest version, no additional steps will be needed. 
Thanks for review! I will try to resolve comments soon.

AKamyshnikova commented on config/crd/patches/cainjection_in_cassandraclusters.yaml:
```diff
@@ -0,0 +1,8 @@
+metadata:
+  annotations:
+    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
+  name: cassandraclusters.db.orange.com
```
I will drop this, this is not enabled via kustomization.yaml so do not effect anything.

AKamyshnikova commented on config/crd/kustomization.yaml:
```diff
@@ -0,0 +1,27 @@
+- bases/db.orange.com_cassandrabackups.yaml
+- bases/db.orange.com_cassandraclusters.yaml
+- bases/db.orange.com_cassandrarestores.yaml
+# +kubebuilder:scaffold:crdkustomizeresource
+
+patchesStrategicMerge:
+# [WEBHOOK] To enable webhook, uncomment all the sections with [WEBHOOK] prefix.
+# patches here are for enabling the conversion webhook for each CRD
+#- patches/webhook_in_cassandrabackups.yaml
+#- patches/webhook_in_cassandraclusters.yaml
+#- patches/webhook_in_cassandrarestores.yaml
+# +kubebuilder:scaffold:crdkustomizewebhookpatch
+
+# [CERTMANAGER] To enable webhook, uncomment all the sections with [CERTMANAGER] prefix.
+# patches here are for enabling the CA injection for each CRD
+#- patches/cainjection_in_cassandrabackups.yaml
+#- patches/cainjection_in_cassandraclusters.yaml
+#- patches/cainjection_in_cassandrarestores.yaml
+# +kubebuilder:scaffold:crdkustomizecainjectionpatch
```
I will remove this. The idea behind is that you can uncomment and got changed configuration.

AKamyshnikova commented on docker/circleci/Dockerfile:
```diff
@@ -54,7 +54,7 @@ RUN curl https://raw.githubusercontent.com/golang/dep/master/install.sh | INSTAL
   tar -xz -C /usr/local/bin gotestsum
 RUN go get -u golang.org/x/lint/golint
-RUN go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.0
+RUN go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.6.2
```
It works but required additionally to rework unittest. As this not requirement for current change, this should be done separatly.

AKamyshnikova commented on main.go:
```diff
@@ -0,0 +1,219 @@
+		return "", fmt.Errorf("%s must be set", watchNamespaceEnvVar)
+	}
+	return ns, nil
+}
```
Yes https://sdk.operatorframework.io/docs/building-operators/golang/migration/#what-was-changed

AKamyshnikova commented on multi-casskop/.gitignore:
```diff
@@ -0,0 +1,24 @@
+
+# Binaries for programs and plugins
+*.exe
+*.exe~
+*.dll
+*.so
+*.dylib
```
No, on macos. I will drop adding gitignore

cscetbon commented:
@AKamyshnikova please try to not force push anymore, it's hard for me to review your last changes. As soon as a review has started you shouldn't force push anymore, that way only the changes can be reviewed and not the whole thing every time.

cscetbon commented:
> Thanks for review! I will try to resolve comments soon.

Let me know when it's ready for review. Also I tried to rerun an old PR tests and the deploy works there. So there must be something wrong on your branch. Take a look at https://app.circleci.com/pipelines/github/Orange-OpenSource/casskop/1840/workflows/ef53e126-ec4b-453c-959d-2d3c20a7b52c/jobs/18369

AKamyshnikova commented:
@cscetbon Sorry, I'm not used to github review process. Always has gerrit review for this :) Will keep that in mind.

AKamyshnikova commented:
Yes, it is ready for review. Can you suggest how I can fix this branch?

fdehay commented:
hello @AKamyshnikova ,
I fixed the circle-ci to github connection issue. It should be ok to run the tests now I think

AKamyshnikova commented:
@fdehay Thank you!

cscetbon commented:
Thanks @AKamyshnikova I'll take care of the few minor things I've seen. We'll try to test it more before making any new release and opening issues if we find some

cscetbon commented:
@AKamyshnikova any idea why in that PR you've renamed the CRD file ? 
```
diff --git a/multi-casskop/deploy/crds/db.orange.com_multicasskops.yaml b/multi-casskop/config/crd/bases/multicluster_v1alpha1_cassandramulticluster_crd.yaml
similarity index 100%
rename from multi-casskop/deploy/crds/db.orange.com_multicasskops.yaml
rename to multi-casskop/config/crd/bases/multicluster_v1alpha1_cassandramulticluster_crd.yaml
```

You might want to create a PR to fix that PR. If yes, I moved the repo to https://github.com/cscetbon/casskop. You can be the 1st to create a PR there. I still need to create a PR here to redirect to my repo as I'm taking over from there.

AKamyshnikova commented:
@cscetbon It is the name that is generated by controller-gen. 
File with such name will always be created, to save old name we can only drop the one that are generated.
Something like this I did in PR with dump to 0.19
https://github.com/Orange-OpenSource/casskop/commit/9c605183474830a2d921fde0923080fe4ba4038b#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R185

cscetbon commented:
But why does it have v1alpha1 in its name ? I'm also fixing an issue in the backup/restore that was introduced by the upgrade of the sdk operator.

AKamyshnikova commented:
I guess because when I start working on PR it was v1alpha1, I send PR with fixes https://github.com/cscetbon/casskop/pull/29
Please point to issues you found and I will try to help.
